auth/indieauth/helpers.go

```diff
@@ -0,0 +1,120 @@
+package indieauth
+
+import (
+	"crypto/sha1" // nolint:gosec
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/andybalholm/cascadia"
+	"github.com/pkg/errors"
+	"github.com/teris-io/shortid"
+	"golang.org/x/net/html"
+)
+
+func createAuthRequest(authDestination, userID, displayName, accessToken, baseServer string) (*Request, error) {
+	authURL, err := url.Parse(authDestination)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to parse IndieAuth destination")
+	}
+
+	authEndpointURL, err := getAuthEndpointFromURL(authURL.String())
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get IndieAuth endpoint from destination URL")
+	}
+
+	baseServerURL, err := url.Parse(baseServer)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to parse local owncast base server URL")
+	}
+
+	callbackURL := *baseServerURL
+	callbackURL.Path = "/api/auth/indieauth/callback"
+
+	codeVerifier := shortid.MustGenerate()
+	codeChallenge := createCodeChallenge(codeVerifier)
+	state := shortid.MustGenerate()
+	responseType := "code"
+	clientID := baseServerURL.String() // Our local URL
+	codeChallengeMethod := "S256"
+
+	redirect := *authEndpointURL
+
+	q := authURL.Query()
+	q.Add("response_type", responseType)
+	q.Add("client_id", clientID)
+	q.Add("state", state)
+	q.Add("code_challenge_method", codeChallengeMethod)
+	q.Add("code_challenge", codeChallenge)
+	q.Add("me", authURL.String())
+	q.Add("redirect_uri", callbackURL.String())
+	redirect.RawQuery = q.Encode()
+
+	return &Request{
+		Me:                 authURL,
+		UserID:             userID,
+		DisplayName:        displayName,
+		CurrentAccessToken: accessToken,
+		Endpoint:           authEndpointURL,
+		ClientID:           baseServer,
+		CodeVerifier:       codeVerifier,
+		CodeChallenge:      codeChallenge,
+		State:              state,
+		Redirect:           &redirect,
+		Callback:           &callbackURL,
+	}, nil
+}
+
+func getAuthEndpointFromURL(urlstring string) (*url.URL, error) {
+	htmlDocScrapeURL, err := url.Parse(urlstring)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to parse URL")
+	}
+
+	r, err := http.Get(htmlDocScrapeURL.String()) // nolint:gosec
+	if err != nil {
+		return nil, err
+	}
+	defer r.Body.Close()
+
+	scrapedHTMLDocument, err := html.Parse(r.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to parse html at remote auth host")
+	}
+	authorizationEndpointTag := cascadia.MustCompile("link[rel=authorization_endpoint]").MatchAll(scrapedHTMLDocument)
```

Comment:
It'd be great if we could support the Discovery endpoint, as although fairly new it's going to be fairly widespread before long https://indieauth.spec.indieweb.org/#indieauth-server-metadata

Reply:
We could also outsource this logic to github.com/hawx/indieauth

Reply:
I apologize in advance for my ignorance, I'm quite detached from the ecosystem and how all these things work together. 
 But I'm confused. There's a not-yet-formalized update to IndieAuth, but a library that hasn't been updated in two years supports it?

I also didn't use an OAuth library because I wasn't building an OAuth implementation, I specifically was following the documentation around IndieAuth.  And not the "new" IndieAuth, the "original" IndieAuth (again I don't really understand the difference, or if they're compatible, etc).

Admittedly I'm a little frustrated by the entire thing and if IndieAuth is going to be an always changing spec that is not going to be used by people who visit Owncast streams I should probably cut my losses because I don't want to cost of maintaining something that's constantly changing but nobody is using.

Reply:
No I completely empathise with you and agree that it is a little frustrating!

Some of the recent changes we've made very recently have been to align much closer to OAuth2, so it's a much more straightforward implementation

And no you're right, that library doesn't yet support the new means, it was something I'm planning on contributing so we'd not need the logic in here either

Reply:
Filed an issue pointing to this thread to possibly work on the next version of IndieAuth if other servers start supporting it and people find the support within Owncast useful. https://github.com/owncast/owncast/issues/1837